src/core_plugins/kibana/public/visualize/listing/visualize_landing_page_table.js

```diff
@@ -162,43 +118,30 @@ export class VisualizeLandingPageTable extends React.Component {
     return this.getColumnSortOrder(TYPE_COLUMN_ID);
   }
 
-  getAreAllItemsSelected() {
-    const { selectedIds, items, currentPage } = this.state;
-    const pageOfItems = this.pager.getItemsOnPage(items, currentPage);
-    const pageOfItemIds = pageOfItems.map(item => item.id);
-    return areAllItemsSelected(selectedIds, pageOfItemIds);
+  areAllItemsSelected() {
+    return areAllItemsSelected(this.state, this.pager);
   }
 
-  getVisualizeColumns() {
+  renderVisualizeColumns() {
     const { selectedIds } = this.state;
     return [
-      getCheckBoxColumn(this.getAreAllItemsSelected(), selectedIds, this.onToggleItem, this.onToggleAll),
-      getTitleColumn(this.getEditUrlForItem, this.getTitleSortOrder(), this.onSortByTitle),
-      getTypeColumn(this.onSortByType, this.getTypeSortOrder()),
+      getCheckBoxColumn(this.areAllItemsSelected(), selectedIds, this.onToggleItem, this.onToggleAll),
+      getTitleColumn(this.getEditUrlForItem, this.getTitleSortOrder(), this.onSort.bind(null, TITLE_COLUMN_ID)),
+      getTypeColumn(this.onSort.bind(null, TYPE_COLUMN_ID), this.getTypeSortOrder()),
     ];
   }
 
-  onDelete = () => {
-    const { deleteItems } = this.props;
-    const { selectedIds, filter } = this.state;
-    deleteItems(selectedIds).then((didDelete) => {
-      if (didDelete) {
-        this.onFilter(filter);
-      }
-    });
-  };
-
-  getActionButtons() {
+  renderActionButtons() {
     return this.state.selectedIds.length > 0
       ? <DeleteButton onClick={this.onDelete} />
       : <CreateButtonLink href={'#' + VisualizeConstants.WIZARD_STEP_1_PAGE_PATH} />;
   }
 
-  getTableContents() {
+  renderTableContents() {
```

Comment:
One of the reasons I went with a column centric approach was because it would be easier to implement moveable columns.  If you wanted to give a header cell functionality to change it's position in the table, this would be difficult with a `bodyRow` react element like that.  In the current column implementation, you just reorder the columns array, and it works.

If you made bodyRow and headerRow an array of cells, you could still rearrange pretty easily but you'd have to do some extra work to make sure they are in sync.

Rearranging columns doesn't really make sense in the landing page tables, but I think it would be useful functionality for the discover table.

What do you think if I modified @weltenwort's suggestion to still use the columns definition like the below?
I still don't love this because then ItemTable can't encapsulate the knowledge of moveable columns, which it could if it was passed just the columns.

```
class DashboardLandingPageTable {
  render() {
    this.columns = getDashboardColumns();
    const headerCells = columns.forEach(column => getHeaderCell());
    const headerRow = <ItemTableHeaderRow> { headerCells }</ItemTableHeaderRow>
    const bodyCells = columns.ForEach(column => getRowCell());
    const bodyRow = <ItemTableBodyRow> { bodyCells } </ItemTableBodyRow>

    return (
      <ItemTable
        bodyRow={bodyRow}
        headerRow={headerRow}
        items={items}
      />
    );
  }
}

function ItemTable({bodyRow, headerRow, items}) {
  return (
    <table class="kuiItemTable">
      <thead>
        { React.cloneElement(headerRow) }
      </thead>
      <tbody>
        { items.map((item) => React.cloneElement(bodyRow, { item })) }
      </tbody>
    </table>
  );
}

function ItemTableBodyRow({ children, item }) {

  if (item.confidential) {
    return (
      <tr class="">
        <td>You do not need to know</td>
      </tr>
    );
  } else {
    return (
      <tr class="tableRow--highlight">
        { React.Children.map(children, (child) => React.cloneElement(child, { item })) }
      </tr>
    );
  }
}

function ItemTableDateCell({ format, item, key }) {
  return (
    <td>
      { formatDate(item[key], format) }
    </td>
  );
}
```

@weltenwort I'm a little confused how this works with the item being passed in though.  Would `ItemTableStringCell` be a react element that takes in an item props and embeds `item.name` inside?  Or does it automatically somehow use the key to fill the inner children?

@cjcenizal I understand your thinking that the row is a single entity, but for a table,  the row cells and the header cells are coupled - it doesn't really make sense for them to exist separately, IMO.  Also think about giving the user the ability to add or hide columns.  That also would be easy given the column centric approach.